Support for "silent authentication/session refresh"

gdestuynder opened the pull request:
Hi,

This is a PR for the code some of Mozilla's RP use - it adds supports for the `prompt` parameter (see http://openid.net/specs/openid-connect-implicit-1_0.html#RequestParameters)

I moved some things around to accommodate. Note also that I initially tried to implement this outside of your library through the `extra_args` parameter, but that just wasn't as nice as having it part of the library.

Additionally, I added code to have the default session expiration based on the requested `id_token` `exp` value (ie do as requested/expected by the OIDC OP) unless the user has this override by settings.
This goes hand in hand with the session refresh, but let me know if you prefer a separate PR.

Finally, this is our reference doc if this helps understanding how/why this work this way: https://wiki.mozilla.org/Security/Guidelines/OpenID_Connect

And a demo of the PR is also running at https://python.testrp.security.allizom.org/ (similar to the example code at the moment)

I'm hoping that you're able to take this PR and find it useful. That said let me know if there is anything I can do/change if necessary to get it included in your code.

Thanks!

gdestuynder commented:
I'll start with fixing these tests actually :)

gdestuynder commented:
turns out the failure has something to do with travis. fixed in https://github.com/mozilla-iam/Flask-pyoidc/pull/2 - will mirror to the PR here as well

gdestuynder commented:
we also ended up prettifying the code, hopefully that's ok with you in one pr

zamzterz commented on README.md:
```diff
@@ -54,6 +54,23 @@ auth = OIDCAuthentication(client_registration_info=client_info)
+
+If your OpenID Connect provider supports the `prompt=none` parameter, the library can automatically support session refresh on your behalf.
+This ensures that the user session attributes (OIDC claims, user being active, etc.) are valid and up-to-date without having to log the user out and back in.
+To use the feature simply pass the parameter requesting the session refresh internval as such:
```
Typo 'internval' -> 'interval'.

zamzterz commented on README.md:
```diff
@@ -54,6 +54,23 @@ auth = OIDCAuthentication(client_registration_info=client_info)
+client_info = {
+    'client_id': 'cl41ekfb9j',
+    'client_secret': 'm1C659wLipXfUUR50jlZ',
+    'session_refresh_interval': 900
```
What is the time unit?
Could this parameter be renamed `session_refresh_interval` -> `session_refresh_interval_seconds` (or what the time unit is) to clarify it?

gdestuynder replied:
seconds indeed

zamzterz commented on README.md:
````diff
@@ -54,6 +54,23 @@ auth = OIDCAuthentication(client_registration_info=client_info)
+auth = OIDCAuthentication(client_registration_info=client_info)
+```
+
+**Note: The client will still be logged out at whichever expiration time you set for the flask session.
````
Casing of Flask: 'flask session' -> 'Flask session'.

zamzterz commented on example/app.py:
```diff
@@ -6,10 +6,21 @@
+app.config.update({'SERVER_NAME': 'example.com',
+                   'SECRET_KEY': 'dev_key',
+                   'PREFERRED_URL_SCHEME': 'https',
+                   'SESSION_PERMANENT': True, #turn on flask session support
```
Missing space after '#' at start of comment.

andrewkrug replied:
Will fix

zamzterz commented on example/app.py:
```diff
@@ -6,10 +6,21 @@
+                   'SECRET_KEY': 'dev_key',
+                   'PREFERRED_URL_SCHEME': 'https',
+                   'SESSION_PERMANENT': True, #turn on flask session support
+                   'PERMANENT_SESSION_LIFETIME': 2592000, #session time in seconds (30 days)
```
Missing space after '#' at start of comment.

andrewkrug replied:
Will fix

zamzterz commented on example/app.py:
```diff
@@ -6,10 +6,21 @@
+                   'PERMANENT_SESSION_LIFETIME': 2592000, #session time in seconds (30 days)
+                   'DEBUG': True})
+
+client_info = {
```
I prefer to keep the example app as simple as possible.
The documentation in the README should be enough for anyone wanting to enable this feature, please remove this added config for the `session_refresh_interval`.

zamzterz commented on src/flask_pyoidc/flask_pyoidc.py:
```diff
@@ -1,81 +1,203 @@
-    def __init__(self, flask_app, client_registration_info=None, issuer=None,
-                 provider_configuration_info=None, userinfo_endpoint_method='POST',
-                 extra_request_args=None):
+    """OIDCAuthentication object for flask extension.
```
Nitpick: casing of 'Flask'.

andrewkrug replied:
Will fix

zamzterz commented on src/flask_pyoidc/flask_pyoidc.py:
```diff
@@ -1,81 +1,203 @@
-                 extra_request_args=None):
+    """OIDCAuthentication object for flask extension.
+
+    Takes a flask app object, client, registration info,
```
Nitpick: casing of 'Flask'.

andrewkrug replied:
Will fix

zamzterz commented on src/flask_pyoidc/flask_pyoidc.py:
```diff
@@ -1,81 +1,203 @@
-        # setup redirect_uri
-        self.app.add_url_rule('/redirect_uri', 'redirect_uri',
-                              self._handle_authentication_response)
+        # setup redirect_uri as a flask route
```
Nitpick: casing of 'Flask'.

andrewkrug replied:
Will fix.

zamzterz commented on src/flask_pyoidc/flask_pyoidc.py:
```diff
@@ -1,81 +1,203 @@
+            self._handle_authentication_response
+        )
+
+        # dynamically add the flask redirect uri to the client info
```
Nitpick: casing of 'Flask'.

andrewkrug replied:
Will fix.

zamzterz commented on src/flask_pyoidc/flask_pyoidc.py:
```diff
@@ -1,81 +1,203 @@
+            return False
+
+    def supports_refresh(self):
+        if ('session_refresh_interval' in self.client_registration_info):
```
This could be simplified to `return 'session_refresh_interval' in self.client_registration_info`.

andrewkrug replied:
Should this block merge?  While this is more concise I don't think it's more readable.

zamzterz replied:
It's a common pattern to collapsing boolean expressions like this, see for example [here](http://cs111.wellesley.edu/~cs111/archive/cs111_spring00/public_html/lectures/boolean-simplification.html).
So I'd argue that it's actually more readable to simplify these kind of structures since it's a common pattern.

zamzterz commented on src/flask_pyoidc/flask_pyoidc.py:
```diff
@@ -1,81 +1,203 @@
+        self.flask_session = flask_session
+        self.client_registration_info = client_registration_info
+
+    def __refresh_time(self):
```
Should use one leading underscore: `__refresh_time` -> `_refresh_time`.

andrewkrug replied:
Will fix.

zamzterz commented on src/flask_pyoidc/flask_pyoidc.py:
```diff
@@ -1,81 +1,203 @@
+
+    """Session object for refresh tokens.
+
+    First class object for comparison of date times during session handling.
```
Could this comment be written without reference to how it's used in the decorator below?
Draft suggestion (which needs some work :)): "Wraps comparison of date time necessary for proper session handling."

andrewkrug replied:
Sure we can change the docstring.

zamzterz commented on src/flask_pyoidc/flask_pyoidc.py:
```diff
@@ -1,81 +1,203 @@
+        expiration itself is set to the id_token expiration so this check is
+        also already taken care of.
+        """
+        if self.flask_session.get('id_token_jwt'):
```
This could probably be simplified to `return self.flask_session.get('id_token_jwt') != None`, right?

andrewkrug replied:
Eventually this should use the oic lib to validate the session again.

zamzterz commented on src/flask_pyoidc/flask_pyoidc.py:
```diff
@@ -1,81 +1,203 @@
+
+    def needs_refresh(self):
+        now = time.time()
+        if self.__refresh_time() < now:
```
This could be simplified to `return self.__refresh_time() < time.time()`.

andrewkrug replied:
Same as other logic collapse. Should this block merge?  While this is more concise I don't think it's more readable.

zamzterz replied:
As above I argue the concise version is more readable. :)

zamzterz commented on src/flask_pyoidc/flask_pyoidc.py:
```diff
@@ -116,55 +252,91 @@ def _handle_authentication_response(self):
+            if current_app.config.get('SESSION_PERMANENT'):
+                flask.session.permanent = True
+                flask.session.permanent_session_lifetime \
+                = id_token.get('exp')-time.time()
```
Spacing around operator, it should be `id_token.get('exp') - time.time()`.

andrewkrug replied:
Will fix

andrewkrug commented:
Re: 80 vs 120 character line limit.  Can reformat to go 120 characters.  Would suggest that to avoid confusion about which pep styles you'd like in the project that you add a flake8 config to this project.  

Is 120 characters the only pep8 rule that you modify?  I'm more than happy to reflow this code.  I just want to avoid three passes through this library for pep8.

zamzterz commented:
That would be great, thanks! :)
The max line length should be the only pep8 rule I modify; I've added it in #18 so you can rebase on master if you want.

gdestuynder commented:
@andrewkrug I fixed the README and `_seconds` in https://github.com/mozilla-iam/Flask-pyoidc/pull/5 - left the others for you as per your comments on you taking them :)

purelogiq commented:
👀

purelogiq commented:
Any hope of this being merged @zamzterz? Would prefer not to fork :)

zamzterz commented:
@purelogiq Do you have a use case so you could help with testing of this?
I'd love some help with that and with finishing up the review comments above.

purelogiq commented:
Yup, I can test this and report my finding + make new branch with those fixes later today or tomorrow if it helps.

zamzterz commented:
That would be awesome, thanks! :-)
I'll try to review your work asap next week (when I'm back from summer
holiday).


On Jul 13, 2017 6:38 PM, "Israel Madueme" <notifications@github.com> wrote:

Yup, I can test this and report my finding + make new branch with those
fixes later today or tomorrow if it helps.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/zamzterz/Flask-pyoidc/pull/17#issuecomment-315133557>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABIP1MVjCT5eUWDNDoB_S2B_h36heK_Zks5sNkgLgaJpZM4Nf4YR>
.

gdestuynder commented:
I've fixed all nits except the 120col reflow and rebased on top of master at https://github.com/mozilla-iam/Flask-pyoidc/pull/6 (this is a force-push due to rebase though so waiting for r+ there before reflecting this here) cc @andrewkrug

zamzterz commented:
Nice work, thanks @gdestuynder! Looking forward to getting this finished and merged :)

gdestuynder commented:
@zamzterz I believe this should be good to go now

zamzterz commented:
Thanks for the awesome work of finishing this 👍 !
